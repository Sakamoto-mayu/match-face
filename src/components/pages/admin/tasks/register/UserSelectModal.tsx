'use client'
import { ReactNode, useState } from "react"
import { createPortal } from "react-dom"
import WhiteButton from "@/components/ui/button/WhiteButton"
import WhiteCheckButton from "@/components/ui/button/WhiteCheckButton"
import Input from "@/components/ui/Input"
import OrangeButton from "@/components/ui/button/OrangeButton"

const teck = ['Java', 'PHP', 'FR', 'CL', 'ML', 'QA']
const state = ['研修中', '待機中', 'アサイン中']
const groupArr = ['', 'グループ1', 'グループ2', 'グループ3', 'グループ4', 'グループ5', 'グループ6', 'グループ7', 'グループ8']

// ラップしたコンポーネントをModal化するコンポーネントを仮作成
const Modal = ({
  children,
  buttonText,
  canCloseByClickingBackground = true
}: {
  children: ReactNode,
  buttonText: string,
  canCloseByClickingBackground?: boolean,
}) => {
  const [isOpened, setIsOpened] = useState(false)

  const open = () => setIsOpened(true)
  const close = () => setIsOpened(false)

  if (!isOpened) {
    return (
      <WhiteButton label={buttonText} onClick={open} className="text-xs" />
    )
  }

  // レンダリングするDOMをbodyに固定するためPortalを使用
  const elmModal = (
    <div className="fixed top-0 left-0 z-20 flex items-center justify-center w-full h-full">
      <div className="relative z-10 px-10 py-6 w-4/5 max-w-3xl overflow-y-auto bg-white translate-y-3">
        {children}
      </div>
      {canCloseByClickingBackground && <div
        className="absolute top-0 left-0 w-full h-full bg-black opacity-30"
        onClick={close}
      />}
    </div>
  )
  return createPortal(elmModal, document.body)
}

// 実際にレンダリングされるモーダルは以下に記述
const UserSelectModal = () => {
  const [ search, setSearch ] = useState('')
  const [ year, setYear ] = useState('')
  const [ month, setMonth ] = useState('')
  const [ group, setGroup ] = useState('')

  // 2000年からの配列を作成
  const maxYear = (new Date()).getFullYear()
  const yearArr = Array(maxYear-1999)
    .fill(2000)
    .map((num, index) => `${num + index}`)
  yearArr.unshift("")
  
  // 月の配列を作成
  const monthArr = Array(12)
    .fill(1)
    .map((num, index) => `${num + index}`)
  monthArr.unshift("")

  return (
    <Modal buttonText="追加">
      <div>
        <h2>▶️ユーザーを選択する</h2>
      </div>
      <div className="flex flex-col item-center border-2 rounded-md w-11/12 mx-auto mt-2 p-8">
        <div className="flex items-center mb-4 ml-12">
          <Input
            id="search"
            type="text"
            value={search}
            onChange={(e) => setSearch(e.target.value)}
            className="border-light-gray text-xs p-1 w-96"
          />
          <WhiteButton label="検索" className="text-xs ml-4 w-16" />
        </div>
        <div className="flex mb-4 ml-12">
          <select
            name="year"
            id="year"
            onChange={(e) => setYear(e.target.value)}
            value={year}
            className="w-16 border-light-gray"
          >
            {yearArr.map((element, index) => (
              <option key={`year_${index}`} value={element}>{element}</option>
            ))}
          </select>
          <span className="ml-2">年</span>
          <select
            name="month"
            id="month"
            onChange={(e) => setMonth(e.target.value)}
            value={month}
            className="w-20 border-light-gray ml-2"
          >
            {monthArr.map((element, index) => (
              <option key={`month_${index}`} value={element}>{element}</option>
            ))}
          </select>
          <span className="ml-2">月入社</span>
        </div>
        <div className="flex items-center mb-4 ml-10">
          {teck.map((element) => (
            <WhiteCheckButton
              key={element}
              label={element}
              className="text-xs w-16 mx-3"
            />
          ))}
        </div>
        <div className="flex items-center mb-4 ml-10">
          {state.map((element) => (
            <WhiteCheckButton
              key={element}
              label={element}
              className="text-xs w-24 mx-3"
            />
          ))}
        </div>
        <div className="ml-12">
          <h2>▶️グループから検索する</h2>
        </div>
        <div className="ml-12">
          <select
            name="group"
            id="group"
            onChange={(e) => setGroup(e.target.value)}
            value={group}
            className="w-96"
          >
            {groupArr.map((element, index) => (
              <option key={`group_${index}`} value={element}>{element}</option>
            ))}
          </select>
        </div>
        <div className="mx-auto mt-8">
          <OrangeButton label="絞り込み" className="w-28 h-8 text-sm" />
        </div>
      </div>
    </Modal>
  )
}

export default UserSelectModal
