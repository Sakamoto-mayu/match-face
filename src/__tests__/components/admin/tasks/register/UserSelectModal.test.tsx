import UserSelectModal from "@/components/pages/admin/tasks/register/UserSelectModal";
import { render, screen, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

// Todo: 非同期関数実装後に自動テストの追加・修正の必要あり

// モーダル展開用関数
const open = () => {
  render(<UserSelectModal />);
  fireEvent.click(screen.getByRole("button"));
}

describe("UserSelectModal.tsx", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    open();
  })

  describe("スナップショットテスト", () => {
    it("モーダルを開いている状態", () => {
      const { container } = render(<UserSelectModal />);
      expect(container).toMatchSnapshot();
    });
  })

  describe("入力テスト", () => {
    const user = userEvent.setup()
    it("検索ボックスの入力テスト", async () => {
      render(<UserSelectModal />);
      const input = screen.getByTestId("search-box") as HTMLInputElement;
      fireEvent.change(input, { target: { value: "田中太郎" } });
      expect(input.value).toBe("田中太郎");
    });
  
    it("入社年の選択テスト", async () => {
      render(<UserSelectModal />);
      // デフォルト選択が空文字かどうか
      const beforeSelect = screen.getByTestId("year:") as HTMLOptionElement;
      expect(beforeSelect.selected).toBe(true);
      // 今年の入社年を選択してテスト
      const thisYear = new Date().getFullYear();
      await user.selectOptions(screen.getByTestId("select-year"), `${thisYear}`);
      const afterSelect = screen.getByTestId(`year:${thisYear}`) as HTMLOptionElement;
      expect(afterSelect.selected).toBe(true);
    });

    it("入社月の選択テスト", async () => {
      render(<UserSelectModal />);
      const beforeSelect = screen.getByTestId("month:") as HTMLOptionElement;
      expect(beforeSelect.selected).toBe(true);
      // 1月を選択してテスト
      await user.selectOptions(screen.getByTestId("select-month"), "1")
      const afterSelect = screen.getByTestId("month:1") as HTMLOptionElement;
      expect(afterSelect.selected).toBe(true);
    })

    // Todo: ユーザーグループは本来fetchで持ってくる情報のため、mock作成の必要あり
    // 今回は仮作成
    it("ユーザーグループの選択テスト", async () => {
      render(<UserSelectModal />);
      const nowGroup = "2023年7月入社";
      const beforeSelect = screen.getByTestId(`group:${nowGroup}`) as HTMLOptionElement;
      expect(beforeSelect.selected).toBe(true);
      // 別のグループを選択
      const newGroup = "2023年7月入社フロント";
      await user.selectOptions(screen.getByTestId("select-group"), newGroup);
      const afterSelect = screen.getByTestId(`group:${newGroup}`) as HTMLOptionElement;
      expect(afterSelect.selected).toBe(true);
    })
  });
  
});
