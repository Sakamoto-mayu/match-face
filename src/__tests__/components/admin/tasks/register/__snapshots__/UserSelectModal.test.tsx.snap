// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserSelectModal.tsx スナップショットテスト モーダルを開いている状態 1`] = `
<div>
  <button
    class="bg-white hover:bg-gray-100 active:bg-gray-200 active:translate-y-0.5 active:shadow-sm py-2 px-4 rounded-full border shadow-md text-xs"
  >
    追加
  </button>
</div>
`;
